tutorials/cli_debugging/debug_legacy.sh

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/bash
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+set -e
+
+# This script builds Holoscan SDK with debug symbols and runs the Endoscopy Tool Tracking application with GDB.
+# Usage: ./debug_legacy.sh [build_type:debug,rel-debug,release]
+
+build_type=${1:-rel-debug}
+holoscan_sdk_tag="v2.2.0"
+holoscan_image="nvcr.io/nvidia/clara-holoscan/holoscan:v2.2.0-dgpu"
+
+SCRIPT_DIR=$(dirname $(realpath $0))
+HOLOHUB_ROOT=$(realpath "${SCRIPT_DIR}/../..")
+tmp_dir=${SCRIPT_DIR}/tmp/cli_debugging
+mkdir -p ${tmp_dir}
+
+# Download and build Holoscan SDK
+pushd ${tmp_dir}
+if [ ! -d holoscan-sdk ]; then
+    git clone git@github.com:nvidia-holoscan/holoscan-sdk.git
+fi
+cd holoscan-sdk
+git checkout ${holoscan_sdk_tag}
+# Apply patch to fix broken Vulkan SDK source
+# https://github.com/nvidia-holoscan/holoscan-sdk/issues/30
+git reset --hard HEAD
+git apply ${SCRIPT_DIR}/holoscan_sdk_20240723_1.diff
+./run build --type $build_type
+INSTALL_DIR=$(realpath $(find . -type d -name "install-*"))
+popd
+
+# Build the tutorial container with GDB
+${HOLOHUB_ROOT}/dev_container build \
+    --img holohub:debugging \
+    --docker_file ${SCRIPT_DIR}/Dockerfile \
+    --base_img ${holoscan_image}
+
+# Build the Endoscopy Tool Tracking application with debugging symbols
+${HOLOHUB_ROOT}/dev_container launch \
+    --docker_opts "-v ${INSTALL_DIR}:/opt/nvidia/holoscan" \
+    --img holohub:debugging \
+    -- bash -c './run build endoscopy_tool_tracking --type ${build_type}'
+
+# Launch GDB with the Endoscopy Tool Tracking application
+${HOLOHUB_ROOT}/dev_container launch \
+    --docker_opts "-v ${INSTALL_DIR}:/opt/nvidia/holoscan --security-opt seccomp=unconfined" \
+    --img holohub:debugging \
+    --  bash -c \
+            'cd /workspace/holohub/build/endoscopy_tool_tracking && \
+            gdb -q \
+                -ex "break main" \
+                -ex "run --data /workspace/holohub/data/endoscopy" \
+                -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
```

Comment:
when I launch the container seperately, holoscan-sdk is not mounted in /workspace. 
Oddly, the call here still works. 
Is there an explanation for this?

Reply:
I'm having trouble reproducing. Here's what I'm seeing:

```
./tutorials/cli_debugging/debug_legacy.sh
# HSDK/container/app build, app launches, then stop container

./dev_container launch --img holohub:debugging

# In the container
cd /workspace/holohub/build/endoscopy_tool_tracking && \
            gdb -q \
                -ex "break main" \
                -ex "run --data /workspace/holohub/data/endoscopy" \
                -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
/workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking
```

With that I see that the application was not able to load Holoscan SDK source file symbols such as for `/workspace/holoscan-sdk/src/core/application.cpp`:
```
Breakpoint 1, 0x00005a3b8fcdee40 in main ()
No source file named /workspace/holoscan-sdk/src/core/application.cpp.
Make breakpoint pending on future shared library load? (y or [n])
```

That behavior is expected, as the v2.2 base container does not include debug symbols with Holoscan SDK binaries in `/opt/nvidia/holoscan`.

Could you please confirm:
- What is the size of your Holoscan SDK binary folder? I would expect to see ~300MB for Release symbols, ~1.1GB for the custom RelWithDebInfo build:
```
$ du -sh /opt/nvidia/holoscan
331M    /opt/nvidia/holoscan
```
- Are you breaking on a `.h` or `.cpp` file? It is expected that you can break on `.h` files under the v2.2 distribution, those are included in the Holoscan SDK binary development package at `/opt/nvidia/holoscan` and symbols can be generated when you build the Endoscopy Tool Tracking app.

Reply:
for relaunching the container I used the following: 
`cd ~/repos/github/holohub/tutorials/cli_debugging
INSTALL_DIR=$(realpath $(find . -type d -name "install-*"))
cd ~/repos/github/holohub
./dev_container launch --docker_opts "-v $INSTALL_DIR:/opt/nvidia/holoscan --security-opt seccomp=unconfined" --img holohub:debugging `

inside the container I then do: 
`cd /workspace/holohub/build/endoscopy_tool_tracking`
`gdb -q \
    -ex "break main" \
    -ex "run --data /workspace/holohub/data/endoscopy" \
    -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow" \
    /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking`
    
    So i do mount the holoscan install location, but its mapped to /opt/nvidia/holoscan. 
    It surprises me that the breakpoint is set for /workspace/holoscna-sdk. 
    `maxofir@6d4a536-lcedt:~/build/endoscopy_tool_tracking$ gdb -q     -ex "break main"     -ex "run --data /workspace/holohub/data/endoscopy"     -ex "break /workspace/holoscan-sdk/src/core/application.cpp:add_flow"     /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking
Reading symbols from /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking...
(No debugging symbols found in /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking)
Breakpoint 1 at 0x13e40
Starting program: /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking --data /workspace/holohub/data/endoscopy
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, 0x0000555555567e40 in main ()
Breakpoint 2 at 0x7ffff6ce6ef0: file /workspace/holoscan-sdk/src/core/application.cpp, line 170.
`

Reply:
Ah, thanks for the extra details, I understand now. Debug symbols are embedded in Holoscan SDK libraries as absolute paths with respect to the build container. That is, interactive debugging will appear to occur with respect to the Holoscan SDK source path at build time, which was mounted at `/workspace/holoscan-sdk`.

That means there are three related but different directories to consider:
- Where the Holoscan SDK source is mounted at build time (`/workspace/holoscan-sdk`) determines debug symbol mappings
- Where the Holoscan SDK binaries are mounted at runtime (`/opt/nvidia/holoscan`) determines what binaries are available
- Where the Holoscan SDK source is mounted at runtime (`/workspace/holoscan-sdk`) has no direct impact on GDB debugging, but is useful for reference

TL;DR: This is expected behavior but can be easily misleading if the Holoscan SDK mount path changes in the run container. I'll add a note in this tutorial warning about that potential confusion.

Reply:
Add a small note in the README on file paths from debugging symbols.